onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc

```diff
@@ -84,6 +85,46 @@ std::unique_ptr<KernelRegistry> RegisterKernels() {
 
 }  // namespace xnnpack
 
+// Not graceful, but we have to define a global variable to hold allocator for all XNNPACK EP instances.
+static AllocatorPtr AllocContext_global{nullptr};
+static std::atomic_int AllocContext_use_count_g{0};
+
+static void* xnn_allocate(void* context, size_t size) {
+  IAllocator* allocator = (*reinterpret_cast<AllocatorPtr*>(context)).get();
+  return allocator->Alloc(size);
```

Comment:
void* to type only requires static_cast IIRC

Can we use IAllocator* for context instead of AllocatorPtr?This code shouldn't need to care if it's in a shared_ptr or a unique_ptr or something else - it just needs an IAllocator.

Reply:
We can store either IAllocatorPtr* or IAllocator** in context as context can be only initialized once. However, `&(AllocatorPtr->get())` looks not good.

Reply:
If we don't use AllocatorPtr(shared_ptr) to own allocator, we can't control its life-cycle.

Reply:
We control the life by the static in init_xnnpack.

Reply:
To keep its maintainbility, I'd make xnn_init and xnn_deinit to be paired, they have the same calling counts.

Sorry, I think I probably make it confused.  You are right, If we would never release AllocatorPtr as long as Process is alive, we can use IAllocator* for context. If I used a referece_count, which means `IAllocator* ` is a variable and can be changed, I have to use `IAllocatorPtr*`.